### PR TITLE
build: migrate from yarn v1 to  yarn berry (v2)

### LINKED ISSUE #558: Migrate to Yarn v2
- [ ] I'd be willing to implement the feature

**Is your feature request related to a problem?**

Verdaccio UI uses Yarn v1. Since this is not a monorepo and we want to learn more about Yarn Berry, we want to run Yarn v2 on this repo from now on. 

**Describe the solution you'd like**

Full migration, with zip files and PnP enabled.

### PR DESCRIPTION
**Pull request by juanpicado**
**Type:** build

**Description:**

For an easy upgrade, we are still using `nodeLinker: node-modules` which is almost the same as `yarn v1`. The idea is move to PnP as soon are issues are fixed. We would need to update webpack and other small issues I found.

> :warning: `security/snyk` fails, this is a known issue, cannot do anything about it but will be fixed eventually. 

- [x] Update CI
- [x]  Migrate to yarn
- [x] Fix test configuration
- [x] Fix lint
- [x] Check build works fine
- [x] Update contribution guidelines
- [x] fix `yarn dev` (it gets hang)
- [x] fix e2e (few test are failing)

 Resolves #558

### CODE REVIEW

File: .yarn/sdks/typescript/lib/tsserver.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+  return Object.assign(Session.prototype, {
+    onMessage(/** @type {string} */ message) {
+      const parsedMessage = JSON.parse(message)
+
+      if (
+        parsedMessage != null &&
+        typeof parsedMessage === 'object' &&
+        parsedMessage.arguments &&
+        parsedMessage.arguments.hostInfo === 'vscode'
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply .yarn/sdks/typescript/lib/tsserver.js:31]: or `isVSCode = parsedMessage?.arguments?.hostInfo === 'vscode'`

File: .yarn/sdks/typescript/lib/tsserver.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    // We add the `zip:` prefix to both `.zip/` paths and virtual paths
+    if (isAbsolute(str) && !str.match(/^\^zip:/) && (str.match(/\.zip\//) || str.match(/\$\$virtual\//))) {
+      // Absolute VSCode `Uri.fsPath`s need to start with a slash.
+      // VSCode only adds it automatically for supported schemes,
+      // so we have to do it manually for the `zip` scheme.
+      // The path needs to start with a caret otherwise VSCode doesn't handle the protocol
+      // https://github.com/microsoft/vscode/issues/105014#issuecomment-686760910
+      return `${isVSCode ? '^' : ''}zip:${str.replace(/^\/?/, `/`)}`;
+    } else {
+      return str;
```
Comment (priscilawebdev) [inline .yarn/sdks/typescript/lib/tsserver.js:58]:
```suggestion
}
      return str;
```

File: .yarn/sdks/typescript/lib/typescript.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env node
+
+const {existsSync} = require(`fs`);
+const {createRequire, createRequireFromPath} = require(`module`);
+const {resolve, dirname} = require(`path`);
+
+const relPnpApiPath = "../../../../.pnp.js";
+
+const absPnpApiPath = resolve(__dirname, relPnpApiPath);
+const absRequire = (createRequire || createRequireFromPath)(absPnpApiPath);
```
Comment (priscilawebdev) [inline .yarn/sdks/typescript/lib/typescript.js:10]: I noticed that we are repeating this in a couple of places... maybe we should make it reusable?

File: babel.config.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+const BROWSER_TARGETS = {
+  browsers: ['last 5 versions', 'FireFox >= 44', 'Safari >= 7', 'Explorer 11', 'last 4 Edge versions'],
```
Comment (priscilawebdev) [inline babel.config.js:2]: at the beginning of  August 17, 2021, Microsoft will no longer support IE 11..should we wait?

File: jest/identity.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable no-var, comma-dangle, no-unused-vars, global-require */
+let Reflect;
+let idObj;
+
+function checkIsNodeV6OrAbove() {
+  if (typeof process === 'undefined') {
```
Comment (priscilawebdev) [inline jest/identity.js:6]:
```suggestion
  if (process === undefined) {
```

File: jest/identity.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable no-var, comma-dangle, no-unused-vars, global-require */
+let Reflect;
```
Comment (priscilawebdev) [inline jest/identity.js:2]: Maybe we should add a description explaining what this file does?

File: package.json [hunk 7]
```diff
@@ -13,18 +13,42 @@
...
+    "@babel/plugin-proposal-decorators": "7.10.5",
+    "@babel/plugin-proposal-export-namespace-from": "7.10.4",
+    "@babel/plugin-proposal-function-sent": "7.10.4",
+    "@babel/plugin-proposal-json-strings": "7.10.4",
     "@babel/plugin-proposal-nullish-coalescing-operator": "7.10.4",
+    "@babel/plugin-proposal-numeric-separator": "7.10.4",
+    "@babel/plugin-proposal-object-rest-spread": "7.11.0",
     "@babel/plugin-proposal-optional-chaining": "7.11.0",
-    "@babel/polyfill": "7.11.5",
+    "@babel/plugin-proposal-throw-expressions": "7.10.4",
```
Comment (priscilawebdev) [inline package.json:28]: Please are these new dependencies necessary for this PR?

File: package.json [hunk 8]
```diff
@@ -166,32 +199,26 @@
...
-    "lint:css": "stylelint \"src/**/styles.ts\"",
-    "lint:lockfile": "lockfile-lint --path yarn.lock --type yarn --validate-https --allowed-hosts verdaccio npm yarn",
+    "lint": "yarn run lint:js && yarn run lint:css",
+    "lint:js": "yarn run type-check && eslint . --ext .js,.ts,.tsx",
+    "lint:html": "yarn htmllinter src/template/index.html",
+    "lint:css": "yarn stylelint \"src/**/styles.ts\"",
     "coverage:publish": "codecov",
-    "pre:webpack": "rimraf static/*",
-    "dev:web": "cross-env BABEL_ENV=ui babel-node tools/dev.server.js",
+    "dev:web": "babel-node tools/dev.server.js",
```
Comment (priscilawebdev) [inline package.json:216]: it looks much more clear :)

File: package.json [hunk 9]
```diff
@@ -219,5 +246,13 @@
     "type": "opencollective",
     "url": "https://opencollective.com/verdaccio",
     "logo": "https://opencollective.com/verdaccio/logo.txt"
+  },
+  "dependenciesMeta": {
```
Comment (priscilawebdev) [inline package.json:250]: Please why this change is necessary?

File: babel.config.js (hunk 4 above)
Comment (priscilawebdev) [reply babel.config.js:2]: Please, does it work? shouldn't it be better to add a browser slug and version?

File: .yarn/sdks/typescript/lib/tsserver.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (juanpicado) [reply .yarn/sdks/typescript/lib/tsserver.js:31]: All inside of `.sdks` is generated code.  https://yarnpkg.com/getting-started/editor-sdks

File: package.json (hunk 7 above)
Comment (juanpicado) [reply package.json:28]: Yes, I removed `@verdaccio/babel-preset`, so the babel dependencies need to be on the root from now on. Yarn 2 is strict on this matter.

File: package.json (hunk 8 above)
Comment (juanpicado) [reply package.json:216]: The reason of those was because `@verdaccio/babel-preset`, so not need it anymore.

File: package.json (hunk 9 above)
Comment (juanpicado) [reply package.json:250]: Only if we use `pnp`, so in this case are not need it, but this command https://yarnpkg.com/cli/unplug#gatsby-focus-wrapper actually create the new `dependenciesMeta`, is a new field introduced by yarn. 

I'll remove it for now.

File: .yarn/sdks/typescript/lib/typescript.js (hunk 3 above)
Comment (juanpicado) [reply .yarn/sdks/typescript/lib/typescript.js:10]: Ignore all inside of `sdks`. I'll add it to `.gitignore`, I think those should be downloaded via vcode.

File: babel.config.js (hunk 4 above)
Comment (juanpicado) [reply babel.config.js:2]: It does work, Babel does it out of the box including few polyfills, like `fetch`. Remove that support won't change much for us to be honest, thus positive impact is almost 0. I'd prefer keep it as long material-ui does it https://material-ui.com/getting-started/supported-platforms/

File: jest/identity.js (hunk 6 above)
Comment (juanpicado) [reply jest/identity.js:2]: I removed one dependency `identity-obj-proxy` and copy the code there, yarn2 did not like it and was breaking all test. https://github.com/verdaccio/ui/issues/93 meant to verify whether still need this, so I might left a comment on the ticket so anyone can match this change.

File: .vscode/extensions.json [hunk 10]
```diff
@@ -0,0 +1,8 @@
+{
+  "recommendations": [
+    "arcanis.vscode-zipfs",
+    "dbaeumer.vscode-eslint",
+    "esbenp.prettier-vscode",
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply .vscode/extensions.json:5]: This will install the `sdks` extensions, so we don't need that code commited.

File: .yarn/sdks/typescript/lib/typescript.js (hunk 3 above)
Comment (merceyz) [reply .yarn/sdks/typescript/lib/typescript.js]: If you're using PnP it should be commited, if not you can just remove it :)

File: .vscode/settings.json [hunk 11]
```diff
@@ -1,15 +1,23 @@
...
+  },
+  "eslint.alwaysShowStatus": true,
+  "search.exclude": {
+    "**/.yarn": true,
+    "**/.pnp.*": true
+  },
+  "eslint.nodePath": ".yarn/sdks",
+  "prettier.prettierPath": ".yarn/sdks/prettier/index.js",
+  "typescript.enablePromptUseWorkspaceTsdk": true,
+  "stylelint.stylelintPath": ".yarn/sdks/stylelint/lib/index.js"
```
Comment (merceyz) [inline .vscode/settings.json:22]: Since you're not using PnP and you removed the SDK you need to remove these otherwise the language server will crash

File: package.json [hunk 12]
```diff
@@ -166,32 +199,26 @@
...
-    "test:clean": "npx jest --clearCache",
+    "test:clean": "jest --clearCache",
     "test:acceptance": "codeceptjs run --steps",
-    "test:acceptance:server": "concurrently --kill-others \"npm run verdaccio:server\" \"npm run test:acceptance\"",
-    "test:e2e": "cross-env BABEL_ENV=test jest --config ./test/jest.config.e2e.js",
-    "test": "cross-env NODE_ENV=test BABEL_ENV=test TZ=UTC jest --config ./jest/jest.config.js --maxWorkers 2 --passWithNoTests",
-    "test:update-snapshot": "npm run test -- -u",
+    "test:acceptance:server": "concurrently --kill-others \"yarn run verdaccio:server\" \"yarn run test:acceptance\"",
+    "test:e2e": "cross-env jest --config ./test/jest.config.e2e.js",
+    "test": "cross-env NODE_ENV=test TZ=UTC jest --config ./jest/jest.config.js --maxWorkers 2 --passWithNoTests",
```
Comment (merceyz) [inline package.json:208]: You don't need `cross-env` with V2, it supports this out of the box

**Review by merceyz — COMMENTED** [review]
Seems the `.yarn/install-state.gz` file snuck past your gitignore https://github.com/verdaccio/ui/pull/572/files#diff-05a728841a7e7287c563b47f8a43e30b

File: package.json (hunk 7 above)
Comment (merceyz) [reply package.json:28]: You can still use the preset but it needs to use `require.resolve` for the plugins/presets it adds

File: .yarn/sdks/typescript/lib/typescript.js (hunk 3 above)
Comment (juanpicado) [reply .yarn/sdks/typescript/lib/typescript.js]: Thanks really appreciated feedback. How this deal with conflicts? I guess eventually the code might be updated.

File: .vscode/settings.json (hunk 11 above)
Comment (juanpicado) [reply .vscode/settings.json:22]: good catch ! thanks

File: .yarn/sdks/typescript/lib/tsserver.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply .yarn/sdks/typescript/lib/tsserver.js]: oh yes :)

**Review by priscilawebdev — APPROVED** [review]
I'd just keep using `@verdaccio/babel-preset ` and use `require.resolve `as @merceyz suggested because I think the dependencies are clearer this way ...but other than that it looks good to me...Good Job 👏

File: .yarn/sdks/typescript/lib/typescript.js (hunk 3 above)
Comment (merceyz) [reply .yarn/sdks/typescript/lib/typescript.js]: When you update Yarn you'd also run `yarn dlx @yarnpkg/pnpify --sdk vscode` to update it, but unless something changes from vscode you most likely wont have to.

File: .github/workflows/canary.yml [hunk 13]
```diff
@@ -17,7 +17,7 @@ jobs:
           node-version: '12.x'
           registry-url: 'https://registry.verdaccio.org'
       - name: Install
-        run: yarn install --frozen-lockfile --network-timeout 100000
+        run: yarn install
```
Comment (merceyz) [inline .github/workflows/canary.yml:20]: `--frozen-lockfile` is now `--immutable` https://yarnpkg.com/cli/install#options

Comment (juanpicado) [reply .github/workflows/canary.yml:20]: thanks ! I did not see this one :)

File: package.json (hunk 7 above)
Comment (juanpicado) [reply package.json:28]: Thanks, on this specific case, I will use explicit dependencies, much clear for contributors what we are using and easier to update or replace.

**Comment by juanpicado** [comment]
Thanks guys for the review :guitar:
